### PR TITLE
Execute FIPS Self Test in Runtime

### PR DESCRIPTION
**Pull request by rusty1968**
This commit enables Runtime to run FIPS Self Tests as follows : KATs, verification of firmware image in ICCM, and ROM integrity test. 

- Light refactor to decouple KAT crate from Rom dependency.
- Light refactor to decouple Image verification crate from ROM dependency.
- Ability for RT to defer execution of self test (basic job control). 

Runtime copies image from ICCM to mailbox SRAM , so it needs to fake a send a transaction and hold the mailbox lock while the image is being validated,. It defers the actual self test execution until the mailbox lock can be acquired. SoC must "poll" for self test completion.

- SoC must send a FIPS self test command one time to schedule the execution.
- Caliptra holds the mailbox lock while the self test is in progress.  
- Failure of FIPS self test is handled as fatal error.
- Upon successful self-test completion the mailbox lock is released and SoC will receive a positive acknowledgement when FIPS Self test command is re-issued.

### CODE REVIEW

File: runtime/src/fips.rs [hunk 1]
```diff
@@ -80,14 +59,47 @@ impl FipsVersionCmd {
     }
 }
 
-pub struct FipsSelfTestCmd;
-impl FipsSelfTestCmd {
+pub mod fips_self_test_cmd {
```
Comment (FerralCoder) [inline runtime/src/fips.rs:62]: nit: this name is now inconsistent with the rest. Either change all or change none.

Comment (rusty1968) [reply runtime/src/fips.rs:62]: this is rust convention for modules.

File: builder/src/lib.rs [hunk 2]
```diff
@@ -57,7 +57,7 @@ pub const FMC_WITH_UART: FwId = FwId {
 pub const APP_WITH_UART: FwId = FwId {
     crate_name: "caliptra-runtime",
     bin_name: "caliptra-runtime",
-    features: &["emu", "test_only_commands"],
+    features: &["emu", "test_only_commands", "fips_self_test"],
```
Comment (jhand2) [inline builder/src/lib.rs:60]: Could you look at how big the `caliptra-runtime` image is with this feature enabled?

I usually just do `readelf -S path/to/caliptra-runtime` and then add the sizes of the `.text` and `.rodata` sections.

File: kat/src/lib.rs [hunk 3]
```diff
@@ -25,8 +26,45 @@ mod sha384acc_kat;
...
+    Sha384AccKat::default().execute(env.sha384_acc)?;
+
+    cprintln!("[kat] ECC-384");
+    Ecc384Kat::default().execute(env.ecc384, env.trng)?;
+
+    cprintln!("[kat] HMAC-384");
+    Hmac384Kat::default().execute(env.hmac384, env.trng)?;
+
+    cprintln!("[kat] LMS");
+    LmsKat::default().execute(env.sha256, env.lms)?;
```
Comment (jhand2) [inline kat/src/lib.rs:65]: Do we need to execute the LMS KAT? We don't use LMS except for to do the ImageVerification (which is also just to satisfy FIPS)

File: runtime/src/fips.rs [hunk 4]
```diff
@@ -80,13 +59,99 @@ impl FipsVersionCmd {
...
+
+        cprintln!("copy manifest");
+        env.mbox.copy_bytes_to_mbox(env.manifest.as_bytes())?;
+
+        cprintln!("copy fmc");
+        let fmc = unsafe { crate::create_slice(FMC_ORG, env.manifest.fmc.size as usize) };
+        env.mbox.copy_bytes_to_mbox(fmc.as_bytes())?;
+        cprintln!("copy rt");
+        let rt = unsafe { crate::create_slice(RUNTIME_ORG, env.manifest.runtime.size as usize) };
+        env.mbox.copy_bytes_to_mbox(rt.as_bytes())?;
```
Comment (jhand2) [inline runtime/src/fips.rs:92]: We could maybe consider putting the rt and fmc regions in the MemoryRegions driver. This would allow rust to enforce accesses better.

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -205,7 +228,21 @@ impl<'a> Drivers<'a> {
...
+    // Run pending jobs before entering low power mode.
+    #[cfg(feature = "fips_self_test")]
+    if let SelfTestStatus::InProgress(execute) = drivers.self_test_status {
+        if drivers.mbox.lock() == false {
+            match execute(drivers) {
+                Ok(_) => drivers.self_test_status = SelfTestStatus::Done,
+                Err(e) => caliptra_drivers::report_fw_error_non_fatal(e.into()),
+            }
+        }
+    }
```
Comment (jhand2) [inline runtime/src/lib.rs:244]: Is there a specific reason SelfTest can't be blocking?

File: runtime/src/lib.rs [hunk 6]
```diff
@@ -255,7 +292,18 @@ fn handle_command(drivers: &mut Drivers) -> CaliptraResult<MboxStatusE> {
...
+        #[cfg(feature = "fips_self_test")]
+        CommandId::SELF_TEST => match drivers.self_test_status {
+            SelfTestStatus::Idle => {
+                drivers.self_test_status = SelfTestStatus::InProgress(fips_self_test_cmd::execute);
+                Ok(MailboxResp::default())
+            }
+            SelfTestStatus::Done => {
+                drivers.self_test_status = SelfTestStatus::Idle;
+                Ok(MailboxResp::default())
+            }
```
Comment (jhand2) [inline runtime/src/lib.rs:304]: I don't really understand this state machine. Can we really serve other commands while verification is running in the background? I would assume that would break our driver access model.

File: kat/src/lib.rs (hunk 3 above)
Comment (rusty1968) [reply kat/src/lib.rs:65]: I see your point. This is  a question for @JohnTraverAmd.

File: runtime/src/lib.rs (hunk 5 above)
Comment (rusty1968) [reply runtime/src/lib.rs:245]: We need to  “reply with data” to conform with new FIPS API (response contains header + checksum), but the mailbox SRAM would already be mutated with the image.

File: runtime/src/lib.rs (hunk 6 above)
Comment (rusty1968) [reply runtime/src/lib.rs:305]: We defer execution to the earliest moment in time when a lock can be acquired.

File: runtime/src/fips.rs (hunk 1 above)
Comment (FerralCoder) [reply runtime/src/fips.rs:63]: agree, but then change them all instead of just changing this one

Comment (FerralCoder) [reply runtime/src/fips.rs:63]: sorry, ignore. The other one I was looking at is a struct, not mod.

File: runtime/src/fips.rs (hunk 4 above)
Comment (rusty1968) [reply runtime/src/fips.rs:92]: I believe the memory regions will be replaced by the persistent data accessor soon. Would it be ok if we defer it to a later PR?

File: builder/src/lib.rs (hunk 2 above)
Comment (rusty1968) [reply builder/src/lib.rs:60]: According to the manifest, the size  of the RT binary image is 48276 bytes.

File: kat/src/lib.rs (hunk 3 above)
Comment (korran) [reply kat/src/lib.rs:65]: LMS is part of the firmware authentication, and fips requires us to reauthenticate. I assume since we do an LMS kat we're actually getting it certified, and thus it's required.

File: runtime/src/lib.rs (hunk 5 above)
Comment (korran) [reply runtime/src/lib.rs:243]: The alternative approach is to make the image-verification digest generic, using the sha384 peripheral instead of sha384acc with the mailbox.

If we want to abandon calling back into ROM forever, perhaps just using the sha384 peripheral is better...

File: runtime/src/fips.rs (hunk 4 above)
Comment (jhand2) [reply runtime/src/fips.rs:92]: Ya, that's ok with me

File: builder/src/lib.rs (hunk 2 above)
Comment (jhand2) [reply builder/src/lib.rs:60]: Thanks!

**Review by FerralCoder — APPROVED** [review]

File: error/src/lib.rs [hunk 7]
```diff
@@ -303,6 +303,7 @@ impl CaliptraError {
     pub const RUNTIME_DISABLE_ATTESTATION_FAILED: CaliptraError =
         CaliptraError::new_const(0x000E0011);
     pub const RUNTIME_HANDOFF_INVALID_PARM: CaliptraError = CaliptraError::new_const(0x000E0012);
+    pub const RUNTIME_SELF_TEST_IN_PROGREESS: CaliptraError = CaliptraError::new_const(0x000E0013);
```
Comment (korran) [inline error/src/lib.rs:306]: typo: Should be RUNTIME_SELF_TEST_IN_PROGRESS

File: kat/src/lib.rs (hunk 3 above)
Comment (jhand2) [reply kat/src/lib.rs:65]: Ok, so we're saying that because we use LMS for firmware auth, we also need to run the LMS KAT. That's fine, just double checking.

File: runtime/src/lib.rs (hunk 5 above)
Comment (korran) [reply runtime/src/lib.rs:243]: I think we should keep this run-self-test mailbox API as is. If a future version of FIPS ever needs us to test the sha384acc peripheral with something other than empty input data we'll need a mechanism to modify the mailbox.

File: test/tests/smoke_test.rs [hunk 8]
```diff
@@ -362,3 +362,82 @@ fn smoke_test() {
...
+    // Confirm we can't re-start the FIPS self test.
+    hw.step_until_boot_status(
+        caliptra_runtime::RtBootStatus::RtFipSelfTestStarted.into(),
+        true,
+    );
+    let _resp = hw
+        .mailbox_execute(u32::from(CommandId::SELF_TEST), payload.as_bytes())
+        .unwrap_err();
+
+    hw.step_until_boot_status(
```
Comment (korran) [inline test/tests/smoke_test.rs:425]: I don't think it is realistic to expect callers to do this. Also, I'm not sure the boot status register is really intended to be used once we've fully booted.

Perhaps the mailbox response from the SELF_TEST command should explain to the user that the test is currently in progress.

Maybe it would be simpler if this were two separate commands `CommandId::SELF_TEST_START` and `CommandID:SELF_TEST_GET_RESULT`?

File: runtime/src/lib.rs [hunk 9]
```diff
@@ -256,7 +291,18 @@ fn handle_command(drivers: &mut Drivers) -> CaliptraResult<MboxStatusE> {
         #[cfg(feature = "test_only_commands")]
         CommandId::TEST_ONLY_HMAC384_VERIFY => HmacVerifyCmd::execute(drivers, cmd_bytes),
         CommandId::VERSION => FipsVersionCmd::execute(drivers),
-        CommandId::SELF_TEST => FipsSelfTestCmd::execute(drivers),
+        #[cfg(feature = "fips_self_test")]
+        CommandId::SELF_TEST => match drivers.self_test_status {
```
Comment (korran) [inline runtime/src/lib.rs:295]: I don't think you should overload the SELF_TEST command for the "start" and "get result" phases. The response in both cases is an empty success result, which isn't very informative. See my comments in smoke_test.rs.

File: runtime/src/lib.rs [hunk 10]
```diff
@@ -256,7 +291,18 @@ fn handle_command(drivers: &mut Drivers) -> CaliptraResult<MboxStatusE> {
...
+            SelfTestStatus::Idle => {
+                drivers.self_test_status = SelfTestStatus::InProgress(fips_self_test_cmd::execute);
```
Comment (jhand2) [inline runtime/src/lib.rs:297]: Does this need to take a function? Or could the handler code in goto_idle just call `fips_self_test_cmd::execute` directly? I feel like the current code is a bit hard to follow.

File: runtime/src/lib.rs [hunk 11]
```diff
@@ -206,7 +229,19 @@ impl<'a> Drivers<'a> {
     }
 }
 
-fn wait_for_cmd(_mbox: &mut Mailbox) {
+/// Run pending jobs and enter low power mode.
+fn goto_idle(drivers: &mut Drivers) {
```
Comment (jhand2) [inline runtime/src/lib.rs:233]: nit: `go_idle`? I think the name `goto_idle` could be misinterpreted as performing a goto jump.

Comment (rusty1968) [reply runtime/src/lib.rs]: Done.

File: test/tests/smoke_test.rs (hunk 8 above)
Comment (rusty1968) [reply test/tests/smoke_test.rs:428]: done

File: runtime/src/lib.rs (hunk 9 above)
Comment (rusty1968) [reply runtime/src/lib.rs]: done

File: runtime/src/fips.rs [hunk 12]
```diff
@@ -78,14 +57,120 @@ impl FipsVersionCmd {
...
+        cprintln!("[rt] Verify complete");
+        Ok(())
+    }
+
+    pub(crate) fn execute(env: &mut Drivers) -> CaliptraResult<()> {
+        caliptra_drivers::report_boot_status(RtFipSelfTestStarted.into());
         cprintln!("[rt] FIPS self test");
-        FipsModule::execute_kats(env)?;
+        copy_and_verify_image(env)?;
+        execute_kats(env)?;
```
Comment (mhatrevi) [inline runtime/src/fips.rs:139]: Shouldn't KATs be executed first before using the cryptographic primitives to verify the image?

**Review by mhatrevi — CHANGES_REQUESTED** [review]

File: runtime/src/fips.rs (hunk 12 above)
Comment (rusty1968) [reply runtime/src/fips.rs:138]: done.

**Comment by rusty1968** [comment]
@mhatrevi , ajisaxena : please give feedback on changes in crates owned by msft.

**Review by mhatrevi — APPROVED** [review]

**Comment by FerralCoder** [comment]
changes were included with PR 693
